File: PVYabause/yabause/YabauseGameCore.m
/*
 Copyright (c) 2009, OpenEmu Team
 

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
     * Redistributions of source code must retain the above copyright
       notice, this list of conditions and the following disclaimer.
     * Redistributions in binary form must reproduce the above copyright
       notice, this list of conditions and the following disclaimer in the
       documentation and/or other materials provided with the distribution.
     * Neither the name of the OpenEmu Team nor the
       names of its contributors may be used to endorse or promote products
       derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY OpenEmu Team ''AS IS'' AND ANY
 EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL OpenEmu Team BE LIABLE FOR ANY
 DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "YabauseGameCore.h"
#import <PVSupport/OERingBuffer.h>
#import <PVSupport/DebugUtils.h>
#import <OpenGLES/gltypes.h>
#import <OpenGLES/ES2/gl.h>
#import <OpenGLES/ES2/glext.h>
#import <PVSupport/PVSupport-Swift.h>

#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include "vdp1.h"
#include "vdp2.h"
//#include "scsp.h"
#include "peripheral.h"
#include "cdbase.h"
#include "yabause.h"
#include "yui.h"
//#include "m68kc68k.h"
#include "cs0.h"
#include "m68kcore.h"
//#include "permacjoy.h"
#include "vidogl.h"
#include "vidsoft.h"

// ToDo: Fix
#define SAMPLERATE 44100
#define SAMPLEFRAME SAMPLERATE / 60

#define HIRES_WIDTH     356
#define HIRES_HEIGHT    256

u32 *videoBuffer = NULL;

int width;
int height;

yabauseinit_struct yinit;
static PerPad_struct *c1, *c2 = NULL;
BOOL firstRun = YES;

// Global variables because the callbacks need to access them...
static OERingBuffer *ringBuffer;

#pragma mark -
#pragma mark OpenEmu SNDCORE for Yabause

#ifndef SNDOE_H
#define SNDOE_H
#define SNDCORE_OE   11
#endif

#define BUFFER_LEN 65536

static int SNDOEInit(void)
{
    return 0;
}

static void SNDOEDeInit(void)
{
}

static int SNDOEReset(void)
{
    return 0;
}

static int SNDOEChangeVideoFormat(UNUSED int vertfreq)
{
    return 0;
}

static void sdlConvert32uto16s(s32 *srcL, s32 *srcR, s16 *dst, u32 len) {
    u32 i;
    
    for (i = 0; i < len; i++)
    {
        // Left Channel
        if (*srcL > 0x7FFF)
            *dst = 0x7FFF;
        else if (*srcL < -0x8000)
            *dst = -0x8000;
        else
            *dst = *srcL;
        srcL++;
        dst++;
        
        // Right Channel
        if (*srcR > 0x7FFF)
            *dst = 0x7FFF;
        else if (*srcR < -0x8000)
            *dst = -0x8000;
        else
            *dst = *srcR;
        srcR++;
        dst++;
    }
}

static void SNDOEUpdateAudio(u32 *leftchanbuffer, u32 *rightchanbuffer, u32 num_samples)
{
    static unsigned char buffer[BUFFER_LEN];
    sdlConvert32uto16s((s32*)leftchanbuffer, (s32*)rightchanbuffer, (s16*)buffer, num_samples);
    
    [ringBuffer write:buffer maxLength:num_samples << 2];
}

static u32 SNDOEGetAudioSpace(void)
{
    return ([ringBuffer usedBytes] >> 2);
}

void SNDOEMuteAudio()
{
}

void SNDOEUnMuteAudio()
{
}

void SNDOESetVolume(UNUSED int volume)
{
}

SoundInterface_struct SNDOE = {
    SNDCORE_OE,
    "Provenance Sound Interface",
    SNDOEInit,
    SNDOEDeInit,
    SNDOEReset,
    SNDOEChangeVideoFormat,
    SNDOEUpdateAudio,
    SNDOEGetAudioSpace,
    SNDOEMuteAudio,
    SNDOEUnMuteAudio,
    SNDOESetVolume
};

#pragma mark -
#pragma mark Yabause Structs

M68K_struct *M68KCoreList[] = {
    &M68KDummy,
    &M68KC68K,
    NULL
};

SH2Interface_struct *SH2CoreList[] = {
    &SH2Interpreter,
    &SH2DebugInterpreter,
    NULL
};

PerInterface_struct *PERCoreList[] = {
    &PERDummy,
//    &PERMacJoy,
    NULL
};

CDInterface *CDCoreList[] = {
    &DummyCD,
    &ISOCD,
    NULL
};

SoundInterface_struct *SNDCoreList[] = {
    &SNDDummy,
    //&SNDMac,
    &SNDOE,
    NULL
};

VideoInterface_struct *VIDCoreList[] = {
    //&VIDDummy,
#ifdef HAVE_LIBGL
    &VIDOGL,
#endif
    &VIDSoft,
    NULL
};

#pragma mark -
#pragma mark OE Core Implementation

@interface YabauseGameCore ()
{
    NSLock  *videoLock;
    
    BOOL    paused;
    NSString *filename;
}

@end

@implementation YabauseGameCore

- (id)init
{
    DLog(@"Yabause init /debug");
    self = [super init];
    if(self != nil)
    {
        filename = [[NSString alloc] init];
        //videoLock = [[NSLock alloc] init];
        videoBuffer = (u32 *)calloc(sizeof(u32), HIRES_WIDTH * HIRES_HEIGHT);
        ringBuffer = [self ringBufferAtIndex:0];
    }
    return self;
}

- (void)setupEmulation
{
    width = HIRES_WIDTH;
    height = HIRES_HEIGHT;
    
    PerPortReset();
    c1 = PerPadAdd(&PORTDATA1);
    c2 = PerPadAdd(&PORTDATA2);
}

- (void)resetEmulation
{
    firstRun = YES;
    YabauseResetButton();
}

- (void)stopEmulation
{
    firstRun = YES;
    [super stopEmulation];
}

- (void)dealloc
{
    YabauseDeInit();
    free(videoBuffer);
}

- (void)initYabauseWithCDCore:(int)cdcore
{
    if ([filename hasSuffix:@".cue"])
    {
        yinit.cdcoretype = CDCORE_ISO;
        yinit.cdpath = [filename UTF8String];
        
        // Get a BIOS
		NSString *bios = [[self BIOSPath] stringByAppendingPathComponent:@"Saturn EU.bin"];
        
        // If a "Saturn EU.bin" BIOS exists, use it otherwise emulate BIOS
        if ([[NSFileManager defaultManager] fileExistsAtPath:bios])
            yinit.biospath = [bios UTF8String];
        else
            yinit.biospath = NULL;
    }
    else
    {
        // Assume we've a BIOS file and we want to run it
        yinit.cdcoretype = CDCORE_DUMMY;
        yinit.biospath = [filename UTF8String];
    }
    
    yinit.percoretype = PERCORE_DEFAULT;
    yinit.sh2coretype = SH2CORE_INTERPRETER;
    
#ifdef HAVE_LIBGL
    yinit.vidcoretype = VIDCORE_OGL;
#else
    yinit.vidcoretype = VIDCORE_SOFT;
#endif
    
    yinit.sndcoretype = SNDCORE_OE;
    yinit.m68kcoretype = M68KCORE_C68K;
    yinit.carttype = CART_DRAM32MBIT; //4MB RAM Expansion Cart
    yinit.regionid = REGION_AUTODETECT;
    
    // Take care of the Battery Save file to make Save State happy
    NSString *path = filename;
    NSString *extensionlessFilename = [[path lastPathComponent] stringByDeletingPathExtension];
    
    NSString *batterySavesDirectory = [self batterySavesPath];
    
    if([batterySavesDirectory length] != 0)
    {
        [[NSFileManager defaultManager] createDirectoryAtPath:batterySavesDirectory withIntermediateDirectories:YES attributes:nil error:NULL];
        NSString *filePath = [batterySavesDirectory stringByAppendingPathComponent:[extensionlessFilename stringByAppendingPathExtension:@"sav"]];
        
        if([filePath length] > 0) {
            DLog(@"BRAM: %@", filePath);
            char *_bramFile;
            const char *tmp = [filePath UTF8String];
            
            _bramFile = (char *)malloc(strlen(tmp) + 1);
            strcpy(_bramFile, tmp);
            yinit.buppath = _bramFile;
        }
    }
    
    yinit.mpegpath = NULL;
    yinit.videoformattype = VIDEOFORMATTYPE_NTSC;
    yinit.frameskip = false;
    yinit.clocksync = 0;
    yinit.basetime = 0;
    yinit.usethreads = 0;
}

- (void)startYabauseEmulation
{
    YabauseInit(&yinit);
    YabauseSetDecilineMode(1);
    OSDChangeCore(OSDCORE_DUMMY);
}

- (CGSize)bufferSize
{
    return CGSizeMake(HIRES_WIDTH, HIRES_HEIGHT);
}

- (CGSize)aspectSize
{
    return CGSizeMake(width, height);
}

- (CGRect)screenRect
{
    return CGRectMake(0, 0, width, height);
}

- (const void *)videoBuffer
{
    return videoBuffer;
}

- (BOOL)loadFileAtPath:(NSString *)path error:(NSError **)error
{
    filename = [path copy];
	DLog(@"Saturn - %@", filename);
    [self setupEmulation];
    [self executeFrame];
    return YES;
}

// Save State is broken, fail more often than not
- (BOOL)saveStateToFileAtPath:(NSString *)path {
#if 0
    ScspMuteAudio(SCSP_MUTE_SYSTEM);
    int error = YabSaveState([fileName UTF8String]);
    ScspUnMuteAudio(SCSP_MUTE_SYSTEM);

    block(!error, nil);
#endif
    return false;
}

- (BOOL)loadStateFromFileAtPath:(NSString *)path
{
#if 0
    ScspMuteAudio(SCSP_MUTE_SYSTEM);
    int error = YabLoadState([fileName UTF8String]);
    ScspUnMuteAudio(SCSP_MUTE_SYSTEM);

    block(!error, nil);
#endif
    return false;
}

#pragma mark -
#pragma mark Inputs

- (oneway void)didPushSaturnButton:(PVSaturnButton)button forPlayer:(NSUInteger)player
{
	PerPad_struct *c = player == 0 ? c1 : c2;
    switch (button)
    {
        case PVSaturnButtonUp:
            PerPadUpPressed(c);
            break;
        case PVSaturnButtonDown:
            PerPadDownPressed(c);
            break;
        case PVSaturnButtonLeft:
            PerPadLeftPressed(c);
            break;
        case PVSaturnButtonRight:
            PerPadRightPressed(c);
            break;
        case PVSaturnButtonStart:
            PerPadStartPressed(c);
            break;
        case PVSaturnButtonL:
            PerPadLTriggerPressed(c);
            break;
        case PVSaturnButtonR:
            PerPadRTriggerPressed(c);
            break;
        case PVSaturnButtonA:
            PerPadAPressed(c);
            break;
        case PVSaturnButtonB:
            PerPadBPressed(c);
            break;
        case PVSaturnButtonC:
            PerPadCPressed(c);
            break;
        case PVSaturnButtonX:
            PerPadXPressed(c);
            break;
        case PVSaturnButtonY:
            PerPadYPressed(c);
            break;
        case PVSaturnButtonZ:
            PerPadZPressed(c);
            break;
        default:
            break;
    }
}

- (oneway void)didReleaseSaturnButton:(PVSaturnButton)button forPlayer:(NSUInteger)player
{
	PerPad_struct *c = player == 0 ? c1 : c2;
    switch (button)
    {
        case PVSaturnButtonUp:
            PerPadUpReleased(c);
            break;
        case PVSaturnButtonDown:
            PerPadDownReleased(c);
            break;
        case PVSaturnButtonLeft:
            PerPadLeftReleased(c);
            break;
        case PVSaturnButtonRight:
            PerPadRightReleased(c);
            break;
        case PVSaturnButtonStart:
            PerPadStartReleased(c);
            break;
        case PVSaturnButtonL:
            PerPadLTriggerReleased(c);
            break;
        case PVSaturnButtonR:
            PerPadRTriggerReleased(c);
            break;
        case PVSaturnButtonA:
            PerPadAReleased(c);
            break;
        case PVSaturnButtonB:
            PerPadBReleased(c);
            break;
        case PVSaturnButtonC:
            PerPadCReleased(c);
            break;
        case PVSaturnButtonX:
            PerPadXReleased(c);
            break;
        case PVSaturnButtonY:
            PerPadYReleased(c);
            break;
        case PVSaturnButtonZ:
            PerPadZReleased(c);
            break;
        default:
            break;
    }
}

#ifdef HAVE_LIBGL
-(BOOL)rendersToOpenGL
{
    return YES;
}
#endif

- (void)executeFrame
{
    if(firstRun) {
        DLog(@"Yabause executeFrame firstRun, lazy init");
        [self initYabauseWithCDCore:CDCORE_DUMMY];
        [self startYabauseEmulation];
        firstRun = NO;
    }
    else {
        //[videoLock lock];
        ScspUnMuteAudio(SCSP_MUTE_SYSTEM);
        YabauseExec();
        ScspMuteAudio(SCSP_MUTE_SYSTEM);
        //[videoLock unlock];
    }
}

- (GLenum)pixelFormat
{
    return GL_RGBA;
}

- (GLenum)internalPixelFormat
{
    return GL_RGBA;
}

- (GLenum)pixelType
{
    return GL_UNSIGNED_BYTE;
}

- (double)audioSampleRate
{
    return SAMPLERATE;
}

- (NSUInteger)channelCount
{
    return 2;
}

#pragma mark -
#pragma mark Yabause Callbacks

void YuiErrorMsg(const char *string)
{
    DLog(@"Yabause Error %@", [NSString stringWithUTF8String:string]);
}

void YuiSwapBuffers(void)
{
    updateCurrentResolution();
    for (int i = 0; i < height; i++) {
        memcpy(videoBuffer + i*HIRES_WIDTH, dispbuffer + i*width, sizeof(u32) * width);
    }
}

#pragma mark -
#pragma mark Helpers

void updateCurrentResolution(void)
{
    int current_width = HIRES_WIDTH;
    int current_height = HIRES_HEIGHT;
    
    // Avoid calling GetGlSize if Dummy/id=0 is selected
    if (VIDCore && VIDCore->id)
    {
        VIDCore->GetGlSize(&current_width,&current_height);
    }
	
    width = current_width;
    height = current_height;
}

@end

File: Cores/Yabause/Classes/YabauseGameCore.swift
//
//  PVYabauseGameCore.swift
//  PVYabause
//
//  Created by Joseph Mattiello on 10/14/18.
//

import Foundation
import PVSupport
import OpenGLES
import Yabause
import PVYabause.Private

@objc
public enum YabauseError : Int, Error {
    case generic
}

@objc
@objcMembers
public class PVYabauseGameCore : PVEmulatorCore, PVSaturnSystemResponderClient {

    static public let USE_THREADS = 0

    // ToDo: Fix
    static public let SAMPLERATE = 44100
    static public let SAMPLEFRAME = SAMPLERATE / 60

    static public let HIRES_WIDTH = 356
    static public let HIRES_HEIGHT = 256

    static public let SNDCORE_OE: Int32 = 11

    public var firstRun = false

    public var width: Int = PVYabauseGameCore.HIRES_WIDTH
    public var height: Int = PVYabauseGameCore.HIRES_HEIGHT

    override public var pixelFormat: GLenum { return GLenum(GL_RGBA) }

    override public var internalPixelFormat: GLenum {
        return GLenum(GL_RGBA)
    }

    override public var pixelType: GLenum {
        return GLenum(GL_UNSIGNED_BYTE)
    }

    override public var audioSampleRate: Double {
        return Double(PVYabauseGameCore.SAMPLERATE)
    }

    override public var channelCount: AVAudioChannelCount {
        return 2
    }

    override public var bufferSize: CGSize {
        return CGSize(width: PVYabauseGameCore.HIRES_WIDTH, height: PVYabauseGameCore.HIRES_HEIGHT)
    }

    override public var aspectSize: CGSize {
        return CGSize(width: width, height: height)
    }

    override public var screenRect: CGRect {
        return CGRect(x: 0, y: 0, width: width, height: height)
    }

    public var videoLock = NSLock()

    public var paused = false
    public var filename : String?

    public var maxDiscs : UInt = 0

    override public func resetEmulation() {
        super.resetEmulation()
        firstRun = true
        YabauseResetButton()
    }

    override public func stopEmulation() {
        firstRun = true
        super.stopEmulation()
    }

    public func startYabauseEmulation() {
        YabauseInit(&yinit)
        YabauseSetDecilineMode(1)
        OSDChangeCore(OSDCORE_DUMMY)
    }

    public func saveStateToFile(atPath path: String) throws {
        var saveError: Int32 = 0
        let lockQueue = DispatchQueue(label: "self")
        lockQueue.sync {
            ScspMuteAudio(SCSP_MUTE_SYSTEM)
            path.withCString {
                saveError = YabSaveState($0)
            }
            ScspUnMuteAudio(SCSP_MUTE_SYSTEM)
        }

        //    NSDictionary *userInfo = @{
        //                               NSLocalizedDescriptionKey: @"Failed to save state.",
        //                               NSLocalizedFailureReasonErrorKey: @"Stella does not support save states.",
        //                               NSLocalizedRecoverySuggestionErrorKey: @"Check for future updates on ticket #753."
        //                               };
        //
        //    NSError *newError = [NSError errorWithDomain:EmulatorCoreErrorCodeDomain
        //                                            code:EmulatorCoreErrorCodeCouldNotSaveState
        //                                        userInfo:userInfo];
        //
        //    *error = newError;
        if saveError != 0 {
            throw YabauseError.generic
        }
    }

    public func loadStateFromFile(atPath path: String) throws {
        var loadError: Int32 = 0

        ScspMuteAudio(SCSP_MUTE_SYSTEM)
        path.withCString {
            loadError = YabLoadState($0)
        }
        ScspUnMuteAudio(SCSP_MUTE_SYSTEM)

        if loadError != 0 {
            throw YabauseError.generic
        }
    }

    override public var supportsSaveStates: Bool {
        return true
    }

    // MARK: -
    // MARK: Inputs
    public func didPush(_ button: PVSaturnButton, forPlayer player: Int) {
        let c = player == 0 ? c1 : c2
        switch button {
        case .up:
            PerPadUpPressed(c)
        case .down:
            PerPadDownPressed(c)
        case .left:
            PerPadLeftPressed(c)
        case .right:
            PerPadRightPressed(c)
        case .start:
            PerPadStartPressed(c)
        case .l:
            PerPadLTriggerPressed(c)
        case .r:
            PerPadRTriggerPressed(c)
        case .a:
            PerPadAPressed(c)
        case .b:
            PerPadBPressed(c)
        case .c:
            PerPadCPressed(c)
        case .x:
            PerPadXPressed(c)
        case .y:
            PerPadYPressed(c)
        case .z:
            PerPadZPressed(c)
        default:
            break
        }
    }

    public func didRelease(_ button: PVSaturnButton, forPlayer player: Int) {
        let c = player == 0 ? c1 : c2
        switch button {
        case .up:
            PerPadUpReleased(c)
        case .down:
            PerPadDownReleased(c)
        case .left:
            PerPadLeftReleased(c)
        case .right:
            PerPadRightReleased(c)
        case .start:
            PerPadStartReleased(c)
        case .l:
            PerPadLTriggerReleased(c)
        case .r:
            PerPadRTriggerReleased(c)
        case .a:
            PerPadAReleased(c)
        case .b:
            PerPadBReleased(c)
        case .c:
            PerPadCReleased(c)
        case .x:
            PerPadXReleased(c)
        case .y:
            PerPadYReleased(c)
        case .z:
            PerPadZReleased(c)
        default:
            break
        }
    }

    #if HAVE_LIBGL
    var rendersToOpenGL: Bool {
        return true
    }
    #endif

    public override func executeFrame() {
        if firstRun {
            DLOG("Yabause executeFrame firstRun, lazy init")
            initYabause(withCDCore: CDCORE_DUMMY)
            startYabauseEmulation()
            firstRun = false
        } else {
            videoLock.lock()
            ScspUnMuteAudio(SCSP_MUTE_SYSTEM)
            YabauseExec()
            ScspMuteAudio(SCSP_MUTE_SYSTEM)
            videoLock.unlock()
        }
    }

    func initYabause(withCDCore cdcore: Int32) {
        guard let filename = filename else {
            // TODO: Throw here
            return
        }
        let `extension` = URL(fileURLWithPath: filename).pathExtension
        let romExtensions = ["cue", "ccd", "mds", "iso"]
        if romExtensions.contains(`extension`), let biosPath = biosPath {
            yinit.cdcoretype = CDCORE_ISO
            filename.withCString {
                yinit.cdpath = $0
            }

            // Get a BIOS
            // sega_101.bin mpr-17933.bin saturn_bios.bin
            let bios = URL(fileURLWithPath: biosPath).appendingPathComponent("saturn_bios.bin").absoluteString

            ILOG("Loading ROM. Will try to setup Saturn BIOS at path: " + bios)

            // If a "Saturn EU.bin" BIOS exists, use it otherwise emulate BIOS
            if FileManager.default.fileExists(atPath: bios) {
                ILOG("BIOS found : " + bios)
                bios.withCString {
                    yinit.biospath = $0
                }
            } else {
                WLOG("No BIOS found")
                yinit.biospath = nil
            }
        } else {
            // Assume we've a BIOS file and we want to run it
            yinit.cdcoretype = CDCORE_DUMMY
            filename.withCString {
                yinit.biospath = $0
            }
        }

        yinit.percoretype = PERCORE_DEFAULT
        yinit.sh2coretype = SH2CORE_INTERPRETER

        #if HAVE_LIBGL
        yinit.vidcoretype = VIDCORE_OGL
        #else
        yinit.vidcoretype = VIDCORE_SOFT
        #endif

        yinit.sndcoretype = PVYabauseGameCore.SNDCORE_OE
        //    yinit.m68kcoretype = M68KCORE_MUSASHI;
        yinit.m68kcoretype = M68KCORE_C68K
        #if true
        yinit.carttype = CART_DRAM32MBIT //4MB RAM Expansion Cart
        #else
        yinit.carttype = CART_NONE
        #endif
        yinit.regionid = u8(REGION_AUTODETECT)
        yinit.buppath = nil
        yinit.mpegpath = nil
        yinit.videoformattype = VIDEOFORMATTYPE_NTSC
        yinit.frameskip = 1
        yinit.clocksync = 0
        yinit.basetime = 0
        #if USE_THREADS
        yinit.usethreads = 1
        yinit.numthreads = 2
        #else
        yinit.usethreads = 0
        #endif

        // Take care of the Battery Save file to make Save State happy
        let path = filename
        let extensionlessFilename = URL(fileURLWithPath: (path as NSString).lastPathComponent).deletingPathExtension().absoluteString

        if let batterySavesDirectory = batterySavesPath {
            try? FileManager.default.createDirectory(atPath: batterySavesDirectory, withIntermediateDirectories: true, attributes: nil)
            let filePath = URL(fileURLWithPath: batterySavesDirectory).appendingPathComponent(URL(fileURLWithPath: extensionlessFilename).appendingPathExtension("sav").absoluteString).absoluteString

            if !filePath.isEmpty {
                DLOG("BRAM: " + filePath)
                filePath.withCString {
                    yinit.buppath = $0
                }
            }
        }
    }

}

#if false
 extension PVYabauseGameCore: DiscSwappable {
    func setMedia(_ inserted: Bool, forDisc disc: UInt) {

    }

    public var numberOfDiscs: UInt {
        return self.numberOfDiscs
    }

    public var currentGameSupportsMultipleDiscs: Bool {
        return numberOfDiscs > 1
    }

    public func swapDisc(number: UInt) {
        setPauseEmulation(false)

        let index = number - 1

        setMedia(true, forDisc: 0)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.setMedia(false, forDisc: index)
        }
    }
 }
#endif
